src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/BaseByteArrayJournalDao.cs

```diff
@@ -156,115 +154,73 @@ private async Task QueueWriteJournalRows(Seq<JournalRow> xs)
-        
-        private async Task HandleDefaultInsert(Seq<JournalRow> xs)
-        {
+            var config = JournalConfig.DaoConfig;
+            
             await using var db = ConnectionFactory.GetConnection();
             await using var tx = await db.BeginTransactionAsync(IsolationLevel.ReadCommitted);
             try
             {
-                await BulkInsertNoTagTableTags(db, xs);
-                await db.CommitTransactionAsync();
-            }
-            catch (Exception e1)
-            {
-                try
+                if (_tagWriteMode == TagWriteMode.Csv)
                 {
-                    await db.RollbackTransactionAsync();
-                }
-                catch (Exception e2)
+                    await db.GetTable<JournalRow>().BulkCopyAsync(new BulkCopyOptions
+                    {
+                        BulkCopyType = xs.Count > config.MaxRowByRowSize ? BulkCopyType.Default : BulkCopyType.MultipleRows,
+                        UseParameters = config.PreferParametersOnMultiRowInsert,
+                        MaxBatchSize = config.DbRoundTripBatchSize
+                    }, xs);
+                } 
+                else
                 {
-                    throw new AggregateException(e2, e1);
+                    // We could not do bulk copy and retrieve the inserted ids
+                    // Issue: https://github.com/linq2db/linq2db/issues/2960
+                    // We're forced to insert the rows one by one.
+                    var tagsToInsert = new List<JournalTagRow>();
+                    foreach (var journalRow in xs)
+                    {
+                        var dbid = await db.InsertWithInt64IdentityAsync(journalRow);
+                        tagsToInsert.AddRange(
+                            journalRow.TagArr.Select(tag => new JournalTagRow
+                            {
+                                JournalOrderingId = dbid, 
+                                TagValue = tag
+                            }));
+                    }
+                    
+                    await db.GetTable<JournalTagRow>().BulkCopyAsync(new BulkCopyOptions
+                    {
+                        BulkCopyType = tagsToInsert.Count > config.MaxRowByRowSize
+                            ? BulkCopyType.Default
+                            : BulkCopyType.MultipleRows,
+                        UseParameters = config.PreferParametersOnMultiRowInsert,
+                        MaxBatchSize = config.DbRoundTripTagBatchSize
+                    }, tagsToInsert);
                 }
                 
-                throw;
-            }
-        }
-
-        private async Task HandleTagTableInsert(Seq<JournalRow> xs)
-        {
-            await using var db = ConnectionFactory.GetConnection();
-            await using var tx = await db.BeginTransactionAsync(IsolationLevel.ReadCommitted);
-            try
-            {
-                await ConsumeSequenceForTagInsert(xs, db);
```

Comment:
So, the purpose of `ConsumeSequenceForTagInsert` was as follows:

Because we are batching all of our writes, but bulk copy is -much- faster than row-by-row, the idea was that in cases of:

`ItemWithTags ItemWithTags ItemNoTags ItemNoTags ItemNoTags ItemNoTags ItemWithTags ItemNoTags` etc... we would still be able to take advantage of bulk-copy for the no-tag cases, and at least lessen the write throughput penalty.